draft-ietf-quic-transport.md

```diff
@@ -3202,27 +3202,32 @@ all subsequent ACK frames containing them could be lost. In this case, the
 loss recovery algorithm could cause spurious retransmits, but the sender will
 continue making forward progress.
 
+
 ### Limiting ACK Ranges {#ack-limiting}
 
-To limit ACK Ranges (see {{ack-ranges}}) to those that have not yet been
-received by the sender, the receiver SHOULD track which ACK frames have been
-acknowledged by its peer. The receiver SHOULD exclude already acknowledged
-packets from future ACK frames whenever these packets would unnecessarily
-contribute to the ACK frame size.  When the receiver is only sending
-non-ack-eliciting packets, it can bundle a PING or other small ack-eliciting
-frame with a fraction of them, such as once per round trip, to enable
-dropping unnecessary ACK ranges and any state for previously sent packets.
-The receiver MUST NOT bundle an ack-eliciting frame, such as a PING, with all
-packets that would otherwise be non-ack-eliciting, in order to avoid an
-infinite feedback loop of acknowledgements.
-
-To limit receiver state or the size of ACK frames, a receiver MAY limit the
-number of ACK Ranges it sends.  A receiver can do this even without receiving
-acknowledgment of its ACK frames, with the knowledge this could cause the sender
-to unnecessarily retransmit some data.  Standard QUIC algorithms
-({{QUIC-RECOVERY}}) declare packets lost after sufficiently newer packets are
-acknowledged.  Therefore, the receiver SHOULD repeatedly acknowledge newly
-received packets in preference to packets received in the past.
+A receiver MAY limit the number of ACK Ranges it sends to limit receiver state
```

Comment:
This normative statement could be misleading.  A receiver really needs to limit the number of ranges somehow, otherwise ACK frames will grow in size indefinitely.  If you do keep it, I think its a SHOULD or MUST not a MAY.

Reply:
Yeah, this might be better as an axiomatic statement.  How about:

> A receiver limits the number of ACK Ranges ({{ack-ranges}}) it remembers and sends in ACK frames, both to limit the size of ACK frames and to avoid resource exhaustion.  A receiver SHOULD stop tracking ACK Ranges after receiving acknowledgments for ACK frames containing ACK Ranges.
>
> It is possible that retaining many ACK Ranges could cause an ACK frame to become too large.  A receiver can discard unacknowledged ACK Ranges to limit ACK frame size.  This is necessary if an ACK frame would be too large to fit in a packet, however receivers MAY also limit ACK frame size further to preserve space for other frames.
>
> When discarding unacknowledged ACK Ranges, a receiver MUST retain the largest acknowledged packet number.  A receiver SHOULD retain ACK Ranges containing newly received packets or higher-numbered packets.

Reply:
One possible addition here, to avoid replay attack:

> A receiver MUST discard packets with packet numbers from ACK Ranges that are not being tracked or they might be exposed to replay attack.  One method for avoiding replay is maintaining a minimum acceptable packet number; see Section 3.4.3 of {{?ESP=RFC4303}} for a sample algorithm.

Reply:
This is a good point, though it might deserve its own PR/Issue?  I thought we had text on dropping packets with packet numbers that had already been processed, but I couldn't find it in transport.

Reply:
It might be in -tls.

Reply:
It's in transport already: https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#section-12.3-11